File: scripts/generate_demo_logs.mjs
import { writeFileSync, mkdirSync } from 'fs'
import { dirname, join } from 'path'
import { fileURLToPath } from 'url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const repoRoot = join(__dirname, '..')

function fmtTs(dt) {
  const pad = (n) => String(n).padStart(2, '0')
  return `${dt.getFullYear()}-${pad(dt.getMonth() + 1)}-${pad(dt.getDate())} ${pad(dt.getHours())}:${pad(dt.getMinutes())}:${pad(dt.getSeconds())}`
}

function mulberry32(seed) {
  return function () {
    let t = (seed += 0x6d2b79f5)
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

function generateSession(startMs, lineCount, intervalSec, tempFn, weights, seed) {
  const rng = mulberry32(seed)
  const events = Object.keys(weights)
  const probs = events.map((e) => weights[e])
  const sum = probs.reduce((a, b) => a + b, 0)
  const lines = []
  let dt = new Date(startMs)

  const pick = () => {
    let r = rng() * sum
    for (let i = 0; i < events.length; i++) {
      r -= probs[i]
      if (r <= 0) return events[i]
    }
    return events[events.length - 1]
  }

  for (let i = 0; i < lineCount; i++) {
    const t = tempFn(i, lineCount, seed + i)
    let kind
    if (i % 6 === 1) kind = 'TEMP_READING'
    else if (i % 6 === 2) kind = 'FAN_SPEED'
    else if (i % 6 === 3) kind = 'HUMIDITY'
    else if (i % 6 === 4) kind = 'VOLTAGE'
    else if (i % 6 === 5) kind = 'BATTERY_LEVEL'
    else kind = pick()

    let event
    if (kind === 'TEMP_READING') event = `TEMP_READING ${t.toFixed(1)}C`
    else if (kind === 'FAN_SPEED') {
      const rpm = Math.max(900, Math.min(3200, Math.round(1400 + (t - 4) * 180 + (rng() * 240 - 120))))
      event = `FAN_SPEED ${rpm}RPM`
    } else if (kind === 'HUMIDITY') event = `HUMIDITY ${Math.round(35 + rng() * 45)}%`
    else if (kind === 'VOLTAGE') {
      const v = 12.4 - (i / lineCount) * 0.6 + (rng() * 0.16 - 0.08)
      event = `VOLTAGE ${v.toFixed(2)}V`
    } else if (kind === 'BATTERY_LEVEL') {
      const b = 99.5 - (i / lineCount) * 18 + (rng() * 0.6 - 0.3)
      event = `BATTERY_LEVEL ${b.toFixed(1)}%`
    } else if (kind === 'SENSOR_TIMEOUT') event = 'SENSOR_TIMEOUT SECONDARY_SENSOR'
    else event = kind

    lines.push(`${fmtTs(dt)} ${event}\n`)
    dt = new Date(dt.getTime() + intervalSec * 1000)
  }
  return lines.join('')
}

function tempStable(i, n, seed) {
  const r = mulberry32(seed + i)
  const base = 4.2 + (i / n) * 2.8
  return Math.max(2.5, Math.min(7.8, base + (r() * 0.8 - 0.4)))
}

function tempExcursion(i, n, seed) {
  const r = mulberry32(seed + i)
  const third = Math.floor(n / 3)
  if (i < third) return 4.5 + r() * 1.5
  if (i < 2 * third) return 7.5 + r() * 2.5
  return 9.5 + r() * 2.5
}

const weightsStable = {
  DOOR_OPEN: 0.08,
  DOOR_CLOSE: 0.08,
  ALARM_TRIGGERED: 0.04,
  TEMP_WARNING: 0.03,
  COOLING_RECOVERY_START: 0.05,
  DEVICE_START: 0.04,
  TELEMETRY_SYNC_FAILED: 0.03,
  SENSOR_TIMEOUT: 0.02,
  TEMP_READING: 0.25,
  FAN_SPEED: 0.12,
  HUMIDITY: 0.1,
  VOLTAGE: 0.08,
  BATTERY_LEVEL: 0.08,
}

const weightsCritical = {
  DOOR_OPEN: 0.12,
  DOOR_CLOSE: 0.1,
  ALARM_TRIGGERED: 0.09,
  TEMP_WARNING: 0.1,
  COOLING_RECOVERY_START: 0.08,
  DEVICE_START: 0.04,
  TELEMETRY_SYNC_FAILED: 0.06,
  SENSOR_TIMEOUT: 0.08,
  TEMP_READING: 0.15,
  FAN_SPEED: 0.08,
  HUMIDITY: 0.06,
  VOLTAGE: 0.07,
  BATTERY_LEVEL: 0.07,
}

const sessions = [
  {
    name: 'medical_device_logs_1000_setB_stable.txt',
    start: new Date('2026-05-20T08:00:00').getTime(),
    tempFn: tempStable,
    weights: weightsStable,
    seed: 42,
  },
  {
    name: 'medical_device_logs_1000_setC_excursion.txt',
    start: new Date('2026-05-22T10:00:00').getTime(),
    tempFn: tempExcursion,
    weights: weightsCritical,
    seed: 77,
  },
]

const outDirs = [
  join(repoRoot, 'docs', 'user-guide', 'examples'),
  String.raw`C:\Users\Daniel-NicolaeVisa\Downloads\OneDrive_1_5-18-2026`,
]

for (const s of sessions) {
  const content = generateSession(s.start, 1000, 9, s.tempFn, s.weights, s.seed)
  for (const dir of outDirs) {
    try {
      mkdirSync(dir, { recursive: true })
      const path = join(dir, s.name)
      writeFileSync(path, content, 'utf8')
      console.log('Wrote', path)
    } catch (e) {
      console.warn('Skip', dir, e.message)
    }
  }
}

File: scripts/generate_demo_logs.py
#!/usr/bin/env python3
"""Generate MED-THERM demo log files (~1000 lines each)."""

from __future__ import annotations

import random
from datetime import datetime, timedelta
from pathlib import Path


def fmt_ts(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%d %H:%M:%S")


def line(dt: datetime, event: str) -> str:
    return f"{fmt_ts(dt)} {event}\n"


def generate_session(
    start: datetime,
    line_count: int,
    interval_seconds: int,
    temp_fn,
    weights: dict[str, float],
    seed: int,
) -> list[str]:
    rng = random.Random(seed)
    events = list(weights.keys())
    probs = [weights[e] for e in events]
    lines: list[str] = []
    dt = start

    for i in range(line_count):
        t = temp_fn(i, line_count)
        # Force telemetry every ~6th line for graphs
        if i % 6 == 1:
            kind = "TEMP_READING"
        elif i % 6 == 2:
            kind = "FAN_SPEED"
        elif i % 6 == 3:
            kind = "HUMIDITY"
        elif i % 6 == 4:
            kind = "VOLTAGE"
        elif i % 6 == 5:
            kind = "BATTERY_LEVEL"
        else:
            kind = rng.choices(events, weights=probs, k=1)[0]

        if kind == "TEMP_READING":
            lines.append(line(dt, f"TEMP_READING {t:.1f}C"))
        elif kind == "FAN_SPEED":
            rpm = int(1400 + (t - 4) * 180 + rng.randint(-120, 120))
            rpm = max(900, min(3200, rpm))
            lines.append(line(dt, f"FAN_SPEED {rpm}RPM"))
        elif kind == "HUMIDITY":
            h = int(35 + rng.randint(0, 45))
            lines.append(line(dt, f"HUMIDITY {h}%"))
        elif kind == "VOLTAGE":
            v = 12.4 - (i / line_count) * 0.6 + rng.uniform(-0.08, 0.08)
            lines.append(line(dt, f"VOLTAGE {v:.2f}V"))
        elif kind == "BATTERY_LEVEL":
            batt = 99.5 - (i / line_count) * 18 + rng.uniform(-0.3, 0.3)
            lines.append(line(dt, f"BATTERY_LEVEL {batt:.1f}%"))
        elif kind == "SENSOR_TIMEOUT":
            lines.append(line(dt, "SENSOR_TIMEOUT SECONDARY_SENSOR"))
        else:
            lines.append(line(dt, kind))

        dt += timedelta(seconds=interval_seconds)

    return lines


# Scenario B: mostly compliant cold chain (~2.5h window)
def temp_stable(i: int, n: int) -> float:
    base = 4.2 + (i / n) * 2.8  # 4.2 -> 7.0
    wobble = random.Random(i).uniform(-0.4, 0.4)
    return max(2.5, min(7.8, base + wobble))


weights_stable = {
    "DOOR_OPEN": 0.08,
    "DOOR_CLOSE": 0.08,
    "ALARM_TRIGGERED": 0.04,
    "TEMP_WARNING": 0.03,
    "COOLING_RECOVERY_START": 0.05,
    "DEVICE_START": 0.04,
    "TELEMETRY_SYNC_FAILED": 0.03,
    "SENSOR_TIMEOUT": 0.02,
    "TEMP_READING": 0.25,
    "FAN_SPEED": 0.12,
    "HUMIDITY": 0.10,
    "VOLTAGE": 0.08,
    "BATTERY_LEVEL": 0.08,
}


# Scenario C: excursion + redundancy loss (~2.5h window)
def temp_excursion(i: int, n: int) -> float:
    third = n // 3
    if i < third:
        return 4.5 + random.Random(i).uniform(0, 1.5)
    if i < 2 * third:
        return 7.5 + random.Random(i).uniform(0, 2.5)
    return 9.5 + random.Random(i).uniform(0, 2.5)  # up to ~12C


weights_critical = {
    "DOOR_OPEN": 0.12,
    "DOOR_CLOSE": 0.10,
    "ALARM_TRIGGERED": 0.09,
    "TEMP_WARNING": 0.10,
    "COOLING_RECOVERY_START": 0.08,
    "DEVICE_START": 0.04,
    "TELEMETRY_SYNC_FAILED": 0.06,
    "SENSOR_TIMEOUT": 0.08,
    "TEMP_READING": 0.15,
    "FAN_SPEED": 0.08,
    "HUMIDITY": 0.06,
    "VOLTAGE": 0.07,
    "BATTERY_LEVEL": 0.07,
}


def main() -> None:
    repo_root = Path(__file__).resolve().parents[1]
    out_dirs = [
        repo_root / "docs" / "user-guide" / "examples",
        Path(r"C:\Users\Daniel-NicolaeVisa\Downloads\OneDrive_1_5-18-2026"),
    ]

    sessions = [
        (
            "medical_device_logs_1000_setB_stable.txt",
            datetime(2026, 5, 20, 8, 0, 0),
            1000,
            9,
            temp_stable,
            weights_stable,
            42,
        ),
        (
            "medical_device_logs_1000_setC_excursion.txt",
            datetime(2026, 5, 22, 10, 0, 0),
            1000,
            9,
            temp_excursion,
            weights_critical,
            77,
        ),
    ]

    for filename, start, count, interval, temp_fn, weights, seed in sessions:
        content = "".join(
            generate_session(start, count, interval, temp_fn, weights, seed)
        )
        for out_dir in out_dirs:
            if not out_dir.exists():
                continue
            path = out_dir / filename
            path.write_text(content, encoding="utf-8")
            print(f"Wrote {path} ({count} lines)")


if __name__ == "__main__":
    main()
